### DISCUSSION TITLE
How do I properly display a growing mutableStateList with Vico and have it display as a Jetpack Compose Preview?

### DISCUSSION THREAD

**Discussion by ericswpark** [patrykandpatrick/vico#795] [Questions] [answered]
### Question

Hi, experimenting with the library and am curious as to whether I'm using the right approach.

I have a viewmodel that has a `mutableStateList` of Kotlin data items. I've created the following composable and preview:

```kotlin
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.tooling.preview.Preview
import com.ericswpark.android.data.SAMPLE_POINT
import com.ericswpark.android.data.SAMPLE_POINT_NO_EXT
import com.ericswpark.android.viewmodel.MeterViewModel
import com.patrykandpatrick.vico.compose.cartesian.CartesianChartHost
import com.patrykandpatrick.vico.compose.cartesian.axis.rememberBottomAxis
import com.patrykandpatrick.vico.compose.cartesian.axis.rememberStartAxis
import com.patrykandpatrick.vico.compose.cartesian.layer.rememberLineCartesianLayer
import com.patrykandpatrick.vico.compose.cartesian.rememberCartesianChart
import com.patrykandpatrick.vico.core.cartesian.data.CartesianChartModelProducer
import com.patrykandpatrick.vico.core.cartesian.data.lineSeries
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext


@Composable
fun MeterGraph(modifier: Modifier = Modifier, meterViewModel: MeterViewModel) {
    val modelProducer = remember { CartesianChartModelProducer() }

    LaunchedEffect(key1 = meterViewModel.dataList.size) {
        withContext(Dispatchers.Default) {
            modelProducer.runTransaction {
                lineSeries { series(meterViewModel.dataList.map { it.getComputedValue() }) }
            }
        }
    }

    CartesianChartHost(
        chart =
        rememberCartesianChart(
            rememberLineCartesianLayer(),
            startAxis = rememberStartAxis(),
            bottomAxis = rememberBottomAxis(),
        ),
        modelProducer = modelProducer,
        modifier = modifier,
    )
}

// TODO: this doesn't work
@Preview(showBackground = true)
@Composable
fun MeterGraphPreview() {
    val meterViewModel = MeterViewModel();
    for (i in 1..10) {
        meterViewModel.addPoint(SAMPLE_POINT)
        meterViewModel.addPoint(SAMPLE_POINT_NO_EXT)
    }

    MeterGraph(
        modifier = Modifier.fillMaxWidth(),
        meterViewModel = MeterViewModel
    )
}
```

Two questions:

1. Is this the right approach to be taking? The parent Composable adds more data to the `dataList` in the view model, and I'm using the key in `LaunchedEffect` to update every time that happens. I think one of the other discussion posts mentioned doing it in the constructor instead, but I couldn't find any documentation regarding this. I'm mainly concerned that this may not be as performant, or not the ideal way to call into the library.

2. The Jetpack Compose preview remains, blank, like so:

![image](https://github.com/user-attachments/assets/1478e4c3-502f-4620-af4c-45b74aabd708)

Is this because `LaunchedEffect` stuff doesn't run in the preview panel? Is there a way to make it show so that I can prototype more quickly? I'm wondering if it's related to feeding in the data using the constructor like mentioned above.

### Vico version(s)

2.0.0-alpha.22

### UI framework(s)

Jetpack Compose

**Comment by Gowsky** [comment, marked as answer]
Hello @ericswpark,

>Is this the right approach to be taking?

If your data is in the `ViewModel`, then instead of having the `CartesianChartModelProducer` in the composable function and running `Transaction`s from a `LaunchedEffect`, it would be preferable to keep the `CartesianChartModelProducer` in the `ViewModel` and feed it with data there. You can then use `MutableStateFlow`, or a similar API instead of `MutableStateList`.

>I think one of the other discussion posts mentioned doing it in the constructor instead, but I couldn't find any documentation regarding this.

It was probably referring to an old version. Using `runTransaction`, as you do, is correct.

>The Jetpack Compose preview remains, blank

Please see #805.
